apps/labs/posts/python-wheels-from-tags-to-variants.md

```diff
@@ -0,0 +1,1083 @@
+---
+
+# Python Wheels: from Tags to Variants
+
+Many Python distributions are uniform across different Python versions
+and platforms. For these distributions, it is sufficient to publish
+a single binary package that can be installed everywhere. However, some
+packages are more complex than that — they include compiled Python
+extensions, binaries or even Python code that differs across systems.
+In order to robustly deploy these software on different platforms,
+you need to publish multiple binary packages, with the installers
+being able to select the one best fit the platform used.
+
+For a long time, [Python Wheels](https://packaging.python.org/en/latest/specifications/binary-distribution-format/)
+made do with a relatively simple mechanism
+of providing the needed variance: [Platform Compatibility Tags](https://packaging.python.org/en/latest/specifications/platform-compatibility-tags/).
+Tags identified different Python implementations and versions,
+operating systems, CPU architectures. Over time, they were extended
+to facilitate new use cases. To list a few: [PEP 513](https://peps.python.org/pep-0513/)
+added <code>manylinux</code> tags to standardize the dependency on GNU/Linux
+systems, [PEP 656](https://peps.python.org/pep-0656/) added
+<code>musllinux</code> tags to facilitate Linux systems with musl libc.
+
+However, not all new use cases could be handled effectively within
+the framework of tags. The advent of GPU-backed computing made distinguishing
+different acceleration frameworks such as CUDA or ROCm important.
+As many distributions have set baselines for their binary packages
+to [x86-64-v2 microarchitecture
+level](https://en.wikipedia.org/wiki/X86-64#Microarchitecture_levels),
+Python packages also started looking at the opportunity
+to express the same requirement. In fact, the [manylinux_2_34 images are
+blocked on x86-64-v2 becoming the default compiler
+target](https://github.com/pypa/manylinux/issues/1585#issuecomment-3094469339).
+Numerical libraries support different
+BLAS/LAPACK, MPI, OpenMP providers — and wish to enable the users to choose
+the build using their desired provider.
+While technically tags could be bent to facilitate all these use cases,
+they would grow quite baroque. Perhaps most importantly, every change
+to tags needs to be implemented in all installers and package-related
+tooling separately, making the adoption difficult.
+
+Facing these limitations, software vendors employed different solutions to work
+around the lack of an appropriate mechanism. Eventually, the [WheelNext](https://wheelnext.dev/)
+took up the challenge to design a more robust solution. For the past 6 months we have
+been working hard on “wheel variants”, in callobration
+with [Astral](https://astral.sh/), [NVIDIA](https://www.nvidia.com/)
+and [PyTorch](https://pytorch.org/) release team. This worked resulted
+in today's [PyTorch 2.8 release](TODO) with experimental variant wheels,
```

Comment:
It's a little confusing to say "variant wheels" here after "wheel variants" above, while I know you're talking about wheels that use variants, I think this might not be clear to readers. You could just say "experimental new wheels"